### PR TITLE
feat: add status/conditions to FeatureTrackers

### PR DESCRIPTION
**Pull request by cam-garrison**
## Description

This PR adds the status.Conditions field to the FeatureTracker object. This is helpful as now we are able to examine FeatureTrackers in the cluster to see if the corresponding Feature was applied successfully.

See this example FeatureTracker for an example of a successful Feature being applied 
```
- apiVersion: features.opendatahub.io/v1
  kind: FeatureTracker
  metadata:
    creationTimestamp: "2024-01-15T18:00:52Z"
    generation: 5
    name: opendatahub-mesh-control-plane-creation
    resourceVersion: "32888"
    uid: 7b71441f-9d6e-41f0-8fd4-64b3ee893319
  spec: {}
  status:
    conditions:
    - lastHeartbeatTime: "2024-01-15T18:01:42Z"
      lastTransitionTime: "2024-01-15T18:00:52Z"
      message: Applying feature mesh-control-plane-creation
      reason: FeatureCreated
      status: "False"
      type: Degraded
    - lastHeartbeatTime: "2024-01-15T18:01:45Z"
      lastTransitionTime: "2024-01-15T18:00:54Z"
      message: Feature mesh-control-plane-creation applied successfully
      reason: FeatureCreated
      status: "True"
      type: Available
```

The phase that the feature is in when it fails/succeeds is indicated in `reason` which can be used in debugging. 

## How Has This Been Tested?

I've added additional unit tests ensuring that conditions are added as expected to FeatureTrackers. 

In additon, I created I tested creating a DSCI with service mesh set to managed and checked that the conditions were added to the featuretrackers - with this image: `quay.io/cgarriso/opendatahub-operator:incubation-add-conditions`

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Review by bartoszmajsak — CHANGES_REQUESTED** [review]

**Review by zdtsw — COMMENTED** [review]
fine with this PR 

saw some comments left there, i can approve once all resolved.

File: pkg/feature/feature.go [hunk 1]
```diff
@@ -253,3 +275,32 @@ func (f *Feature) createResourceTracker() error {
 
 	return nil
 }
+
+func (f *Feature) UpdateFeatureTrackerStatus(condType conditionsv1.ConditionType, status corev1.ConditionStatus, reason featurev1.FeaturePhase, message string) {
```
Comment (bartoszmajsak) [inline pkg/feature/feature.go:279]: I think we need to rework this one a little bit. It's not `nil` safe. Let's wait for #812 and do it together.

File: tests/integration/features/features_int_test.go [hunk 2]
```diff
@@ -127,6 +132,74 @@ var _ = Describe("preconditions", func() {
...
+				}).
+				Load()
+			Expect(err).ToNot(HaveOccurred())
+
+			// when
+			Expect(verificationFeature.Apply()).ToNot(Succeed())
+
+			// then
+			featureTracker := getFeatureTracker("default-post-condition-failure")
+			Expect(featureTracker.Status.Conditions).To(HaveCondition(conditionsv1.ConditionDegraded, v1.ConditionTrue, featurev1.PostConditions))
```
Comment (bartoszmajsak) [inline tests/integration/features/features_int_test.go:198]: This is a bit cryptic, let's improve it.

File: tests/integration/features/features_suite_int_test.go [hunk 3]
```diff
@@ -71,6 +72,9 @@ var _ = BeforeSuite(func() {
 	Expect(err).NotTo(HaveOccurred())
 	Expect(config).NotTo(BeNil())
 
+	err = featurev1.AddToScheme(testScheme)
+	Expect(err).NotTo(HaveOccurred())
```
Comment (bartoszmajsak) [inline tests/integration/features/features_suite_int_test.go:76]: We should move it next to L56 and make it look like:

```golang
	utilruntime.Must(featurev1.AddToScheme(testScheme))
```

Comment (cam-garrison) [reply tests/integration/features/features_suite_int_test.go]: [9455151](https://github.com/opendatahub-io/opendatahub-operator/pull/801/commits/94551518ce359dce2f402dc0b865accad940aeba)

File: tests/integration/features/features_int_test.go (hunk 2 above)
Comment (cam-garrison) [reply tests/integration/features/features_int_test.go]: [9455151](https://github.com/opendatahub-io/opendatahub-operator/pull/801/commits/94551518ce359dce2f402dc0b865accad940aeba)

File: pkg/feature/feature.go (hunk 1 above)
Comment (cam-garrison) [reply pkg/feature/feature.go]: [bc7d71e](https://github.com/opendatahub-io/opendatahub-operator/pull/801/commits/bc7d71e3c202b3a69f5ca72e60b637411418ea29)

File: pkg/feature/feature.go [hunk 4]
```diff
@@ -46,17 +50,30 @@ func newFeature(name string) *Feature {
...
-	if err := f.createResourceTracker(); err != nil {
-		return err
+	if trackerErr := f.createFeatureTracker(); err != nil {
+		return trackerErr
 	}
 
 	// Verify all precondition and collect errors
 	var multiErr *multierror.Error
+	var phase featurev1.FeaturePhase
+	phase = featurev1.FeatureCreated
```
Comment (bartoszmajsak) [inline pkg/feature/feature.go:66]:
```suggestion
	phase := featurev1.FeatureCreated
```

File: tests/integration/features/features_int_test.go [hunk 5]
```diff
@@ -202,28 +207,91 @@ var _ = Describe("feature cleanup", func() {
...
+			// when
+			Expect(verificationFeature.Apply()).To(Succeed())
+
+			// then
+			featureTracker := getFeatureTracker("default-crd-verification")
+			Expect(*featureTracker.Status.Conditions).To(ContainElement(
+				MatchFields(IgnoreExtras, Fields{
+					"Type":   Equal(conditionsv1.ConditionAvailable),
+					"Status": Equal(v1.ConditionTrue),
+					"Reason": Equal(string(featurev1.FeatureCreated)), // Cast to string for testing.
```
Comment (bartoszmajsak) [inline tests/integration/features/features_int_test.go:240]:
```suggestion
					"Reason": Equal(string(featurev1.FeatureCreated)),
```

File: tests/integration/features/features_int_test.go [hunk 6]
```diff
@@ -202,28 +207,91 @@ var _ = Describe("feature cleanup", func() {
...
+			// when
+			Expect(verificationFeature.Apply()).ToNot(Succeed())
+
+			// then
+			featureTracker := getFeatureTracker("default-crd-verification")
+			Expect(*featureTracker.Status.Conditions).To(ContainElement(
+				MatchFields(IgnoreExtras, Fields{
+					"Type":   Equal(conditionsv1.ConditionDegraded),
+					"Status": Equal(v1.ConditionTrue),
+					"Reason": Equal(string(featurev1.PreConditions)), // Cast to string for testing.
```
Comment (bartoszmajsak) [inline tests/integration/features/features_int_test.go:264]:
```suggestion
					"Reason": Equal(string(featurev1.PreConditions)),
```

File: tests/integration/features/features_int_test.go [hunk 7]
```diff
@@ -202,28 +207,91 @@ var _ = Describe("feature cleanup", func() {
...
+			// when
+			Expect(verificationFeature.Apply()).ToNot(Succeed())
+
+			// then
+			featureTracker := getFeatureTracker("default-post-condition-failure")
+			Expect(*featureTracker.Status.Conditions).To(ContainElement(
+				MatchFields(IgnoreExtras, Fields{
+					"Type":   Equal(conditionsv1.ConditionDegraded),
+					"Status": Equal(v1.ConditionTrue),
+					"Reason": Equal(string(featurev1.PostConditions)), // Cast to string for testing.
```
Comment (bartoszmajsak) [inline tests/integration/features/features_int_test.go:289]:
```suggestion
					"Reason": Equal(string(featurev1.PostConditions)),
```

**Review by bartoszmajsak — APPROVED** [review]
/lgtm

File: pkg/feature/feature.go [hunk 8]
```diff
@@ -46,17 +50,29 @@ func newFeature(name string) *Feature {
...
-	if err := f.createResourceTracker(); err != nil {
-		return err
+	if trackerErr := f.createFeatureTracker(); err != nil {
+		return trackerErr
 	}
 
 	// Verify all precondition and collect errors
 	var multiErr *multierror.Error
+	phase := featurev1.FeatureCreated
+	f.updateFeatureTrackerStatus(conditionsv1.ConditionDegraded, "False", phase, fmt.Sprintf("Applying feature %s", f.Name))
```
Comment (zdtsw) [inline pkg/feature/feature.go:66]: why here is set to conditionsv1.ConditionDegraded , 
should not be ConditionProgressing ?

Comment (bartoszmajsak) [reply pkg/feature/feature.go:66]: Good spot. Lost between the lines.

```suggestion
	f.updateFeatureTrackerStatus(conditionsv1.ConditionProgressing, "False", phase, fmt.Sprintf("Applying feature %s", f.Name))
```

**Comment by zdtsw** [comment]
anywhere in the code, we still use this one https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/pkg/gvr/gvr.go#L18 ?

**Comment by zdtsw** [comment]
i think you will need to add one more here https://github.com/cam-garrison/opendatahub-operator/blob/incubation-add-conditions/controllers/dscinitialization/dscinitialization_controller.go#L71C1-L71C128
as 
// `+kubebuilder:rbac:groups="features.opendatahub.io",resources=featuretrackers/status,verbs=get;update;patch;delete`

**Comment by zdtsw** [comment]
> i think you will need to add one more here [cam-garrison/opendatahub-operator@`incubation-add-conditions`/controllers/dscinitialization/dscinitialization_controller.go#L71C1-L71C128](https://github.com/cam-garrison/opendatahub-operator/blob/incubation-add-conditions/controllers/dscinitialization/dscinitialization_controller.go?rgh-link-date=2024-01-24T17%3A05%3A51Z#L71C1-L71C128) as // `+kubebuilder:rbac:groups="features.opendatahub.io",resources=featuretrackers/status,verbs=get;update;patch;delete`

for reference:
024-01-24T16:50:40Z	ERROR	features	Error updating FeatureTracker status	{"feature": "mesh-control-plane-creation", "error": "featuretrackers.features.opendatahub.io \"opendatahub-mesh-control-plane-creation\" is forbidden: User \"system:serviceaccount:openshift-operators:opendatahub-operator-controller-manager\" cannot update resource \"featuretrackers/status\" in API group \"features.opendatahub.io\" at the cluster scope"}
github.com/opendatahub-io/opendatahub-operator/v2/pkg/feature.(*Feature).updateFeatureTrackerStatus
	/workspace/pkg/feature/feature.go:237
github.com/opendatahub-io/opendatahub-operator/v2/pkg/feature.(*Feature).Apply
	/workspace/pkg/feature/feature.go:66
github.com/opendatahub-io/opendatahub-operator/v2/pkg/feature.(*FeaturesInitializer).Apply
	/workspace/pkg/feature/initializer.go:35
github.com/opendatahub-io/opendatahub-operator/v2/controllers/dscinitialization.(*DSCInitializationReconciler).configureServiceMesh
	/workspace/controllers/dscinitialization/servicemesh_setup.go:26
github.com/opendatahub-io/opendatahub-operator/v2/controllers/dscinitialization.(*DSCInitializationReconciler).Reconcile
	/workspace/controllers/dscinitialization/dscinitialization_controller.go:275
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
	/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:122
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:323
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:274
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:235

**Comment by cam-garrison** [comment]
> i think you will need to add one more here https://github.com/cam-garrison/opendatahub-operator/blob/incubation-add-conditions/controllers/dscinitialization/dscinitialization_controller.go#L71C1-L71C128 as // `+kubebuilder:rbac:groups="features.opendatahub.io",resources=featuretrackers/status,verbs=get;update;patch;delete`

thanks for pointing this out - added the fix and tested in cluster to ensure it was fixed. Also fixed some merge conflicts

**Review by zdtsw — APPROVED** [review]

**Comment by zdtsw** [comment]
not related to your PR but during my testing ( no OSSM installed, CRD exists, creating DSCI) I would expect one FT "mesh-control-plane-creation" gets created and failure "FeaturePreConditions"
but obviously, by only checking if CRD existing or not to indicate `EnsureServiceMeshOperatorInstalled`() success or not, thats is not enough. 
i.e in this case, it is meaningless to create another FT "mesh-metrics-collection" and mark "mesh-control-plane-creation" failed with "ApplyManifests"

**Comment by bartoszmajsak** [comment]
> not related to your PR but during my testing ( no OSSM installed, CRD exists, creating DSCI) I would expect one FT "mesh-control-plane-creation" gets created and failure "FeaturePreConditions"
but obviously, by only checking if CRD existing or not to indicate EnsureServiceMeshOperatorInstalled() success or not, thats is not enough.
i.e in this case, it is meaningless to create another FT "mesh-metrics-collection" and mark "mesh-control-plane-creation" failed with "ApplyManifests"

@zdtsw At this point we are not able to define an order of features - as in a chain of dependent features. I think that's what you are looking for.

We simply apply one after the other assuming it will eventually be consistent. I am not sure if I get it right, but what you are expecting is to have only one FT created for `mesh-control-plane-creation` but the other one should not be even triggered. If that's desired behaviour we need to enhance the Feature builder so we can define dependent features. Shouldn't be a big deal.

**Pull request merged**